File: src/main/java/com/milaboratory/mixcr/cli/CommandExportClonesPretty.java
/*
 * Copyright (c) 2014-2022, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.cli;

import cc.redberry.primitives.Filter;
import com.milaboratory.core.alignment.MultiAlignmentHelper;
import com.milaboratory.core.sequence.NSequenceWithQuality;
import com.milaboratory.core.sequence.NucleotideSequence;
import com.milaboratory.mixcr.basictypes.*;
import gnu.trove.set.hash.TLongHashSet;
import io.repseq.core.Chains;
import io.repseq.core.GeneFeature;
import io.repseq.core.GeneType;
import picocli.CommandLine.Command;
import picocli.CommandLine.Option;

import java.io.BufferedOutputStream;
import java.io.FileOutputStream;
import java.io.PrintStream;
import java.util.ArrayList;
import java.util.List;

import static cc.redberry.primitives.FilterUtil.ACCEPT_ALL;
import static cc.redberry.primitives.FilterUtil.and;

@Command(name = "exportClonesPretty",
        sortOptions = true,
        separator = " ",
        description = "Export verbose information about clones.")
public class CommandExportClonesPretty extends ACommandSimpleExportMiXCR {
    public static final int LINE_LENGTH = 80;
    public static final int LINE_OFFSET = 7;
    public static final int MAX_LENGTH = 2 * LINE_OFFSET + LINE_LENGTH;

    @Option(description = "Limit number of alignments before filtering",
            names = {"-b", "--limitBefore"})
    public Integer limitBefore = null;

    @Option(description = "Limit number of filtered alignments; no more " +
            "than N alignments will be outputted",
            names = {"-n", "--limit"})
    public Integer limitAfter = null;

    @Option(description = "List of clone ids to export",
            names = {"-i", "--clone-ids"})
    public List<Long> ids = new ArrayList<>();

    @Option(description = "Number of output alignments to skip",
            names = {"-s", "--skip"})
    public Integer skipAfter = null;

    @Option(description = "Filter export to a specific protein chain gene (e.g. TRA or IGH).",
            names = {"-c", "--chains"})
    public String chain = "ALL";

    @Option(description = "Only output clones where CDR3 (not whole clonal sequence) exactly equals to given sequence",
            names = {"-e", "--cdr3-equals"})
    public String cdr3Equals = null;

    @Option(description = "Only output clones where target clonal sequence contains sub-sequence.",
            names = {"-r", "--clonal-sequence-contains"})
    public String csContain = null;

    public Chains getChain() {
        return Chains.parse(chain);
    }

    TLongHashSet getCloneIds() {
        if (ids.isEmpty())
            return null;
        return new TLongHashSet(ids);
    }

    @SuppressWarnings("unchecked")
    public Filter<Clone> mkFilter() {
        final Chains chains = getChain();

        List<Filter<Clone>> filters = new ArrayList<>();

        final TLongHashSet cloneIds = getCloneIds();
        if (cloneIds != null)
            filters.add(object -> cloneIds.contains(object.getId()));

        filters.add(object -> {
            for (GeneType gt : GeneType.VJC_REFERENCE) {
                VDJCHit bestHit = object.getBestHit(gt);
                if (bestHit != null && chains.intersects(bestHit.getGene().getChains()))
                    return true;
            }
            return false;
        });

        if (csContain != null) {
            csContain = csContain.toUpperCase();
            filters.add(object -> {
                for (int i = 0; i < object.numberOfTargets(); i++)
                    if (object.getTarget(i).getSequence().toString().contains(csContain))
                        return true;
                return false;
            });
        }

        if (cdr3Equals != null) {
            final NucleotideSequence seq = new NucleotideSequence(cdr3Equals);
            filters.add(object -> {
                NSequenceWithQuality feature = object.getFeature(GeneFeature.CDR3);
                return feature != null && feature.getSequence().equals(seq);
            });
        }

        if (filters.isEmpty())
            return ACCEPT_ALL;

        return and(filters.toArray(new Filter[filters.size()]));
    }

    @Override
    public void run0() throws Exception {
        Filter<Clone> filter = mkFilter();
        long total = 0, filtered = 0;
        CloneSet cloneSet = CloneSetIO.read(in);

        try (PrintStream output = out == null ? System.out :
                new PrintStream(new BufferedOutputStream(new FileOutputStream(out), 32768))) {
            long countBefore = limitBefore == null ? Long.MAX_VALUE : limitBefore;
            long countAfter = limitAfter == null ? Long.MAX_VALUE : limitAfter;
            long skipAfter = this.skipAfter == null ? 0 : this.skipAfter;
            for (final Clone clone : cloneSet) {
                if (--countBefore < 0)
                    break;

                ++total;

                if (filter.accept(clone)) {
                    ++filtered;

                    if (--skipAfter >= 0)
                        continue;

                    if (--countAfter < 0)
                        break;

                    outputCompact(output, clone);
                }
            }
            output.println("Filtered: " + filtered + " / " + total + " = " + (100.0 * filtered / total) + "%");
        }
    }

    public static void outputCompact(PrintStream output, final Clone clone) {
        output.println(">>> Clone id: " + clone.getId());
        output.println(">>> Abundance, reads (fraction): " + clone.getCount() + " (" + clone.getFraction() + ")");
        output.println();
        for (int i = 0; i < clone.numberOfTargets(); i++) {
            MultiAlignmentHelper targetAsMultiAlignment = VDJCAlignmentsFormatter.getTargetAsMultiAlignment(clone, i, true, false);
            if (targetAsMultiAlignment == null)
                continue;
            MultiAlignmentHelper[] split = targetAsMultiAlignment.split(80);
            for (MultiAlignmentHelper spl : split) {
                output.println(spl);
                output.println();
            }
        }
        output.println();
    }

}

File: src/main/kotlin/com/milaboratory/mixcr/cli/CommandExportClonesPretty.kt
/*
 * Copyright (c) 2014-2022, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.cli

import cc.redberry.primitives.Filter
import com.milaboratory.core.sequence.NucleotideSequence
import com.milaboratory.mixcr.basictypes.Clone
import com.milaboratory.mixcr.basictypes.CloneSetIO
import com.milaboratory.mixcr.basictypes.MultiAlignmentFormatter
import com.milaboratory.mixcr.basictypes.VDJCAlignmentsFormatter
import com.milaboratory.mixcr.cli.CommonDescriptions.Labels
import com.milaboratory.mixcr.util.and
import gnu.trove.set.hash.TIntHashSet
import io.repseq.core.Chains
import io.repseq.core.GeneFeature
import io.repseq.core.GeneType
import picocli.CommandLine.Command
import picocli.CommandLine.Help.Visibility.ALWAYS
import picocli.CommandLine.Option
import picocli.CommandLine.Parameters
import java.io.BufferedOutputStream
import java.io.FileOutputStream
import java.io.PrintStream
import java.nio.file.Path
import java.util.*

@Command(
    description = ["Export verbose information about clones."]
)
class CommandExportClonesPretty : MiXCRCommandWithOutputs() {
    @Parameters(
        description = ["Path to input file with clones"],
        index = "0",
        paramLabel = "clones.(clns|clna)"
    )
    lateinit var input: Path

    @Parameters(
        index = "1",
        paramLabel = "output.txt",
        arity = "0..1",
        description = ["Path where to write export. Will write to output if omitted."]
    )
    var out: Path? = null

    @Option(
        description = ["Limit number of alignments before filtering"],
        names = ["-b", "--limitBefore"],
        paramLabel = "<n>",
        order = OptionsOrder.main + 10_100
    )
    var limitBefore: Int? = null

    @Option(
        description = ["Limit number of filtered alignments; no more than N alignments will be outputted"],
        names = ["-n", "--limit"],
        paramLabel = "<n>",
        order = OptionsOrder.main + 10_200
    )
    var limitAfter: Int? = null

    @Option(
        description = ["Number of output alignments to skip"],
        names = ["-s", "--skip"],
        paramLabel = "<n>",
        order = OptionsOrder.main + 10_300
    )
    var skipAfter: Int? = null

    @Option(
        description = ["List of clone ids to export"],
        names = ["-i", "--clone-ids"],
        paramLabel = "<id>",
        order = OptionsOrder.main + 10_400
    )
    var ids: List<Int> = mutableListOf()

    @Option(
        description = ["Filter export to a specific protein chain gene (e.g. TRA or IGH)."],
        names = ["-c", "--chains"],
        paramLabel = Labels.CHAINS,
        showDefaultValue = ALWAYS,
        order = OptionsOrder.main + 10_500
    )
    var chains: Chains = Chains.ALL

    @Option(
        description = ["Only output clones where CDR3 (not whole clonal sequence) exactly equals to given sequence"],
        names = ["-e", "--cdr3-equals"],
        paramLabel = "<seq>",
        order = OptionsOrder.main + 10_600
    )
    var cdr3Equals: NucleotideSequence? = null

    @Option(
        description = ["Only output clones where target clonal sequence contains sub-sequence."],
        names = ["-r", "--clonal-sequence-contains"],
        paramLabel = "<seq>",
        order = OptionsOrder.main + 10_700
    )
    var csContain: String? = null

    override val inputFiles
        get() = listOf(input)

    override val outputFiles
        get() = listOfNotNull(out)

    private val cloneIds: TIntHashSet?
        get() = if (ids.isEmpty()) null else TIntHashSet(ids)

    private fun mkFilter(): Filter<Clone> {
        var resultFilter: Filter<Clone> = Filter { true }
        val cloneIds = cloneIds
        if (cloneIds != null) {
            resultFilter = resultFilter.and { clone: Clone -> clone.id in cloneIds }
        }
        resultFilter = resultFilter.and { clone: Clone ->
            GeneType.VJC_REFERENCE
                .mapNotNull { clone.getBestHit(it) }
                .any { bestHit -> chains.intersects(bestHit.gene.chains) }
        }
        if (csContain != null) {
            val csContain = csContain!!.uppercase(Locale.getDefault())
            resultFilter = resultFilter.and { clone: Clone ->
                (0 until clone.numberOfTargets())
                    .map { i -> clone.getTarget(i).sequence }
                    .any { sequence ->
                        sequence.toString().contains(csContain)
                    }
            }
        }
        if (cdr3Equals != null) {
            resultFilter = resultFilter.and { clone: Clone ->
                clone.getFeature(GeneFeature.CDR3)?.sequence == cdr3Equals
            }
        }
        return resultFilter
    }

    override fun validate() {
        ValidationException.requireFileType(input, InputFileType.CLNX)
    }

    override fun run0() {
        val filter = mkFilter()
        var total: Long = 0
        var filtered: Long = 0
        val cloneSet = CloneSetIO.read(input)
        (out?.let { PrintStream(BufferedOutputStream(FileOutputStream(it.toFile()), 32768)) }
            ?: System.out).use { output ->
            val countBefore = limitBefore ?: Int.MAX_VALUE
            val countAfter = limitAfter ?: Int.MAX_VALUE
            val skipAfter = skipAfter ?: 0
            cloneSet
                .asSequence()
                .take(countBefore)
                .onEach { ++total }
                .filter { filter.accept(it) }
                .drop(skipAfter)
                .take(countAfter)
                .forEach { clone ->
                    ++filtered
                    outputCompact(output, clone)
                }

            output.println("Filtered: " + filtered + " / " + total + " = " + 100.0 * filtered / total + "%")
        }
    }

    companion object {
        @JvmStatic
        fun outputCompact(output: PrintStream, clone: Clone) {
            output.println(">>> Clone id: " + clone.id)
            output.println(">>> Abundance, reads (fraction): " + clone.count + " (" + clone.fraction + ")")
            output.println()
            for (i in 0 until clone.numberOfTargets()) {
                val targetAsMultiAlignment = VDJCAlignmentsFormatter(addReads = false)
                    .formatMultiAlignments(clone, i)
                val split = targetAsMultiAlignment.split(80)
                for (spl in split) {
                    output.println(spl.format(MultiAlignmentFormatter.LinesFormatter(addHitScore = true)))
                    output.println()
                }
            }
            output.println()
        }
    }
}
